src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,497 @@
+
+/* Iteration parameters, smaller step and larger iterations is more accurate but
+ * consumes more cycles.
+ */
+#define AZ_STEP			Q_CONVERT_FLOAT(0.3927, 12)	/* PI/16 as Q4.12 */
+#define AZ_ITERATIONS		5				/* loops in min err search */
+#define SOURCE_DISTANCE		Q_CONVERT_FLOAT(5.0, 12)	/* source distance in m Q4.12 */
+
+/* Sound direction angle filtering */
+#define SLOW_AZ_C1		Q_CONVERT_FLOAT(0.01, 15)
+#define SLOW_AZ_C2		Q_CONVERT_FLOAT(1 - 0.01, 15)
+
+/* Threshold for notifying user space, no more often than every 100 ms */
+#define CONTROL_UPDATE_MIN_TIME	Q_CONVERT_FLOAT(100e-3, 16)
+
+/* Emphasis filter for sound direction */
+uint32_t iir_emphasis_48k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xd7b49168, 0x62c27f22,
+	0x19712b75, 0xcd1da916, 0x19712b75, 0x00000000,
+	0x00007fb2
+};
+
+uint32_t iir_emphasis_16k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xef646e59, 0x2cc88203,
+	0x0fb60984, 0xe093ecf7, 0x0fb60984, 0x00000000,
+	0x00007fb2
+};
+
+/* Function to copy data from component source to cross correlation buffer with emphasis
+ * filtering. Skip if not tracking.
+ */
+void tdfb_direction_copy_emphasis(struct tdfb_comp_data *cd, int channels, int *channel, int32_t x)
+{
+	if (!cd->direction_updates)
+		return;
+
+	*cd->direction.wp = iir_df2t(&cd->direction.emphasis[*channel], x) >> 16;
+	(*channel)++;
+	cd->direction.wp++;
+	tdfb_cinc_s16(&cd->direction.wp, cd->direction.d_end, cd->direction.d_size);
+	if (*channel == channels)
+		*channel = 0;
+}
+
+static inline int16_t max_mic_distance(struct tdfb_comp_data *cd)
+{
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int32_t d2;
+	int32_t d2_max = 0;
+	int i;
+	int j;
+
+	/* Max lag based on largest array dimension. Microphone coordinates are Q4.12 meters */
+#ifdef DEBUG
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		printf("printf: Mic%d (%d, %d, %d)\n", i, cd->mic_locations[i].x,
+		       cd->mic_locations[i].y, cd->mic_locations[i].z);
+	}
+#endif
+
+	for (i = 0; i < cd->config->num_mic_locations; i++) {
+		for (j = 0; i < cd->config->num_mic_locations; i++) {
+			if (j == i)
+				continue;
+
+			dx = cd->mic_locations[i].x - cd->mic_locations[j].x;
+			dy = cd->mic_locations[i].y - cd->mic_locations[j].y;
+			dz = cd->mic_locations[i].z - cd->mic_locations[j].z;
+
+			/* d2 is Q8.24 meters */
+			d2 = dx * dx + dy * dy + dz * dz;
+			if (d2 > d2_max)
+				d2_max = d2;
+		}
+	}
+
+	/* z = pow(x, y), where x is Q6.26, y is Q2.30, z is Q12.20, the shifts below
+	 * return the max array dimension as meters Q4.12;
+	 */
+	return drc_pow_fixed(d2_max << 2, ZERO_POINT_FIVE) >> 8;
+}
+
+static inline bool line_array_mode_check(struct tdfb_comp_data *cd)
+{
+	int32_t prod;
+	int i;
+	int num_mic_locations = cd->config->num_mic_locations;
+
+	if (num_mic_locations == 2)
+		return true;
+
+	/* TODO expand check to (x,y,z), this is just (x,y) */
```

Comment:
still todo ?

Reply:
Thanks for reminder. Also in this version not done, into next PR update.

Reply:
Done!